### ISSUE TITLE
API semantic versioning in Apicurio

### ISSUE THREAD

**Issue by KevinMitchell** [Apicurio/apicurio-registry#1112]
I’m a little confused about how versioning is supposed to work with Apicurio Registry.  Each of our APIs uses semantic versioning, and this version is tracked in the Info object of our Open API specifications.  The version in Apicurio Registry.is simply an increasing integer for each artifact.  So what’s the recommended way of mapping APIs to Apicurio artifacts?  Does each API version get mapped to its own artefact, and so you’d typically have one Apicurio version.  Because if the API changed in any way the semantic version would also be altered and you’d need a new artifact?  Or is the suggestion we ignore the semantic versioning concept entirely when using Apicurio?  Or create a new artifact for each new major version of an API, map the minor versions to Apicurio versions, and not use patch versions?  Given the artifacts have versions that seem to be ignored by Apicurio it might be worth adding a section in the documentation describing the recommended approach(es) to versioning as it’s not entirely obvious how we are supposed to map versioned APIs to Apicurio.

**Comment by laundry-96** [comment]
I'm confused on this as well. I dug into the source code and it seems that whenever a file is uploaded to the artifactID, it increments the version, based on how many artifact versions there are. It is irrelevant to what the actual version of what your API is.

What I'm thinking is that someone will have to modify the code to print out the API version in the drop down, instead of the "version" that the Database is using.

**Comment by EricWittmann** [comment]
This is a great question/discussion.  First I'll explain what we have today, then maybe we can discuss potential improvements for the future. :)

You are correct that the version is just an increasing integer for each artifact version.  This concept works fine for the original use-case of e.g. Avro or Protobuf schemas, which don't tend to use semantic versioning the way API designs do.  So the current concept of "version" in Registry does not map well onto API versioning.

Right now, I like the option of mapping a Major version of an API to an artifact in the registry and then only using minor versions without patch.  That's restrictive and not ideal, but it should work OK I think.

Now, the question is how can we make this a lot better?  I think there is an opportunity to allow the version # to be either specified by the user via the API (when creating/updating) or extracted from the content (when possible).  In this case, the artifact version would actually map directly to the semantic version.  Also, we could update the logic of the Compatibility rule so that it actually understood the rules of semantic versioning, where it could apply different rules depending on the version number.

Thoughts @Apicurio/developers ?

**Comment by KevinMitchell** [comment]
Making a new entity for each major version has the advantage that all versions of an entity should be backwards compatible as they will only differ by the minor version.  However, this only partially helps.  Because if we are simply encoding the major version in the entity name then we’ll end up with people using different conventions for these names making it hard to enforce consistency.  And if we want to search for the latest version of an API we’d have to know the major version name encoding for the API.  So I think simply encoding the major version in the entity name isn’t sufficient.  

Furthermore for API specifications with version information in the specification itself it’s important this information is kept consistent with the metadata stored in Apicurio.  We shouldn’t be allowed to upload a new specification with version 1.2.3 into an artifact representing major version 2 of the API for example.  So I think Apicurio needs to know about semantic versioning in such cases.  And it may then be cleaner to have a single artifact representing all semantic versions, even if it means making the “backwards compatibility” checks a little more refined.  I.e. a new minor version update for version 2 should be backwards compatible with the last minor version for major version 2, but instances of the entity for different major versions are ignored.

Of course we then have to reconcile the current Apicurio versioning approach with the semantic version.  The current approach has the advantage that every new upload generates a new version.  Which avoids the question of what should happen if we upload multiple versions of an API with the same semantic version.  Do they overwrite each other, do subsequent uploads fail forcing the minor version in the spec to be incremented before uploading etc.  For API specs it seems common to just use major and minor versions; it’s hard to make sense of a patch version involving an “internal” change for an API spec.  So maybe you could ignore any patch version in the spec and replace it by Apicurio’s version?  That way when uploading an entity you’d have the same behaviour as currently, just incrementing the (patch) version, but with a semantic version prefix?  Or if you wanted to keep the full semantic version from the spec then add the Apicurio version as a build version suffix?  

Of course once you’ve decided how the different versions are stored and versioned then you’d still need to add support for querying such entities.  E.g. what major versions of an API exist, what’s the latest minor version of an API for a given major version etc.  But hopefully they would be easy to implement once the basic approach has been decided on.

**Comment by bobmorley** [comment]
We decided to use semantic versioning for my JSON schemas and now that I am digging in further, have run into some related problems. To ensure robustness, we thought it was reasonable to have someone provide 1.0.0 followed by 1.0.2 and then someone else "fill in the gap" with 1.0.1. Clearly this doesn't work, breaking the UI dropdown ordering and API endpoints like "GetLatest" which is returning "Last uploaded".

This use case might seem edgy, but it seems to me that one could have a situation where 1.x.x is replaced by 2.0.0, and then further development causes a patch/minor increase on the original 1.x.x branch of versions. Certainly for API spec.

**Comment by EricWittmann** [comment]
Yeah if we start working on a semantic versioning feature, then all of these types of issues start to come into play.  We'll need to address sort order, get latest, auto-detection of next version #, etc.  I think it's a super interesting area that would have a ton of value if we can do it right.

We've not started on anything like this yet.  If someone wanted to contribute, even if just by documenting the requirements, that would be super helpful!  Otherwise this is one that we'll just have to get to when we get to...
